[DR] Result paging strategy for related entities query (#908)
OpenAssetIO/OpenAssetIO: doc/decisions/DR016-One-to-many-result-paging.md

- Manager plugin must implement yet another interface.
- Complexity of binding to multiple target languages.
- A mutable, stateful cursor violates the stateless design principle of
  interactions with manager plugins. This causes complications, e.g.
  thread-safety.
- An immutable cursor object (such that `.next()` returns a new cursor
  object), could increase optimisation complexity for the manager
  implementation, e.g. managing query cache sharing, lifetime and
  updates across multiple "live" cursors.

## Considering batching

Additional considerations arise when considering how batching can be
handled algongside paging. There are three broad strategies

- Split batching from paging, using a different API method for each.
- Page through the results from all the batched input elements in
  lockstep, i.e. whole-batch paging.
- Page through the results of each batched input element independently,
  using individual sub-queries after the initial whole-batch query,
  i.e. per-element paging.

Each of these batched paging designs combines more or less well with the
various paging options presented above.

Henceforth the offset-based solution is rejected, given the lack of

> flexibility afforded to the manager's implementation.

Reviewer: [fix] I'd be good to call out that we also reject it for other reasons too (lack of stablity, potential overhead of record counting)

Author: Updated the text in ac493901

  - [W3C IndexedDB Cursor API spec](https://w3c.github.io/IndexedDB/#idbcursor)

### Rejected options

In order to avoid considering a long list of options that are either
clearly inferior, or broadly similar to other options, some valid
solutions were rejected early.

#### Whole-batch offset

The classic offset-based solution was rejected due to the critical
drawback of unstable paging, which a page token based solution overcomes
without introducing much more complexity.

```python
SuccessCallback = Callable[[int, List[EntityReferences]], None]

def relatedEntities(
        batch: List[EntityReference],
        relationship,
        successCb: SuccessCallback,
        errorCb,
        limit: int = -1, offset: int = 0): ...
```

##### Pros

- Conceptually simple for the host.
- Jumping arbitrarily to nth page is trivial.
- Retains batch-awareness for each page.
- Transactions, indexing and caching technologies exist that can
  alleviate stability and performance problems, to some extent.

##### Cons

- Anomalous page results if insertions/deletions can happen before next
  page is retrieved, or if ordering is not stable.

> - Server cannot negotiate page size - there is no mechanism to signal

Author: Added page size negotiation Con after offline conversation (also corresponding Pro for page token) in 042c17a5

  functions or arguments.
- Function invocations to fetch subsequent pages must pass all the same
  arguments, such that they are valid for the page token. This is
  error-prone and/or redundant.

#### Option 1.1 - Whole-batch page token

Host supplies a batch of entity references and a callback. The callback
is called for each input entity, with one of its arguments being a
page of related entities as a simple list.

Once callbacks have been called, the function returns a page token. This
can be supplied to a subsequent call to get the next page of results for
all entities in the input batch simultaneously.

Note that this could also be designed to use a cursor object with
similar pros and cons, but that option was discarded (see Background
section).

```python
class AbstractPageToken:
    ...

# Callback taking index of batch element and list of related entities.
SuccessCallback = Callable[[int, List[EntityReferences]], None]

def relatedEntities(
        batch: List[EntityReference],
        relationship,
        successCb: SuccessCallback,
        errorCb,
        limit: int = -1,
        pageToken: AbstractPageToken = None) -> AbstractPageToken: ...
```

```python

> # Usage:

Author: Added usage examples (here and for other options) after offline conversation in e9e343ab

Reviewer: Don't worry about spending too much time on the batch, paged illustrations,  as we know were going to exclude them for now, def need to document manager implementation sketches for token vs cursor in singular form.

- Potential difficulty for some hosts/queries to determine a priori
  which API (paging or batch) is the most optimal choice.
- Having the choice adds a foot-gun for host developers to shoot
  themselves with by accident.
- Precludes hosts atomically expressing more complex queries that may
  be required by spreadsheet-type UIs.

### Batched paging option 2 - Whole-batch paging

E.g. Page token

```python
# Callback taking index of batch element and selected page of results.
SuccessCallback = Callable[[int, List[EntityReference]], None]

def relatedEntities(
    batch: List[EntityReference], relationship,
    successCb: SuccessCallback, errorCb,
    limit=-1, pageToken=None
) -> AbstractPageToken: ...
```

E.g. Cursor object

```python
class AbstractPage:
    def get(self) -> List[List[EntityReference]]:
        ...
    def next(self, pageSize):
        ...

def relatedEntities(
    batch: List[EntityReference], relationship) -> AbstractPage: ...
```

#### Pros

- Whole batch is paged at once, retaining batch-awareness in the
  manager for subsequent pages.
- Conceptually simpler API for both host and manager if only a single
  page is active across the whole batch.

> ### Pros (page token specific)

Reviewer: [fix]
```suggestion
#### Pros (page token specific)
```

Author: Fixed in e8ca9f26

- Per-element callback means it's trivial to skip processing elements of
  the input batch whose pages are exhausted.
- Retains existing design in line with OpenAssetIO design principles,
  i.e. non-paged signature is simply the same function but ignoring
  optional arguments.

#### Cons

> - Wasted processing of empty lists by the host when pages are exhausted

Reviewer: [fix] I think this needs to go to cursor specific?

Author: Good spot, fixed in a4c2b516

  i.e. when there are still pages remaining for some inputs but not
  others.

> - Batch elements cannot enter an error state in isolation, since paging

Reviewer: [question] Is this only true if the cursor returns a list of lists, vs calling the callback for each element as before? Would probably be good to keep the callback mechanism.

Author: It's tricky.  If one of the batch elements enters an error state in the page token design, then what's the correct way to get the next page of results for the remaining batch elements?  

You need to provide _some_ page token for the erroring element.  Perhaps the last one that was successful (if any were)?  Perhaps a `None` page token can mean "ignore"?  Perhaps you just have to deal with the error callback being called again and again for that failing batch element.

In any case, it's not really possible to consider them erroring in isolation.

Reviewer: Got you - might be good to elaborate in the con a little to explain in more detail

  is whole-batch.
- Collaborators working on manager plugins raised this as being
  particularly tricky to implement.

#### Cons (page token specific)

> - Likely that the manager plugin must combine multiple back-end page

Reviewer: [question] This is true of the cursor approach too right? Just they're aggregated in the cursor object not the token object? Or is this really touching on the topic of is a token a value or an object?

Author: :thinking: It came from considering REST APIs that use page tokens and assuming that ultimately (whether in the plugin or some back-end) the batch request will actually be several requests with different page tokens, so a single uber page token must be a concatenation of some kind.

But like you say, if that was the back-end behaviour then wrapping in a cursor object doesn't solve the problem.

I'll "promote" that Con to a general problem with whole-batch paging.

Author: Moved in cef989e7

  tokens into one, adding complexity to ensure they are maintained in
  sync.

#### Cons (cursor object specific)

- Additional metadata in the object is limited to whole-batch data.

### Batched paging option 3 - Per-element paging

E.g. Page token

```python
# Callback taking index of batch element and selected page of results.
SuccessCallback = Callable[[int, List[EntityReference]], None]

def relatedEntities(
    batch: List[EntityReference],
    relationship,
    successCb: SuccessCallback,
    errorCb,
    limit: -1,
    pageTokens: List[AbstractPageToken]
): -> List[AbstractPageToken]: ...
```

E.g. Cursor object

```python
class AbstractPage:
    def get(self) -> List[EntityReferences]:
        ...
    def next(self, pageSize):
        ...

# Callback taking index of batch element and cursor object.
SuccessCallback = Callable[[int, AbstractPage], None]

def relatedEntities(
    batch: List[EntityReference],
    relationship,
    successCb: SuccessCallback,
    errorCb): ...
```

#### Pros

> - Manager is under no obligation to consider errors in other streams

Reviewer: [question]
```suggestion
- Manager is under no obligation to consider errors in other batch elements
```

Author: Re-worded in 7df1b89b

  when implementing query behaviour, as pagination method need not be
  synchronized.

#### Pros (cursor object specific)

- Individual cursors can be incremented independently, potentially
  reducing wasted queries.

#### Cons

- Batching is very difficult to optimize for the manager, especially
  after the first page.
- Batch-optimized queries have unpredictable performance profile for
  the host, given the wide range of potential implementations.

### Outcome

Given the reported difficulty of crafting queries that are both batch
and paging-aware, we opt to go with Option 1 and split the paging API(s)
from the batched API(s).

This then leaves only two remaining solutions: (non-batched) page token
or cursor objects.

## Considering manager implementations

### Using Python's DB-API standard

> #### Batch signature

Reviewer: [question] The preceding sentence rules out batching, be good to remove or move to the end?

Author: Batching isn't ruled out, instead it is split out from paging, so there's a separate sketch that considers batching only without paging.

Reviewer: > ... we opt to go with Option 1 and split the paging API(s)
from the batched API(s).
>
> This then leaves only two remaining solutions: (non-batched) page token
or cursor objects.

Reads as excluded!

Author: Ah OK, I'll update that text.

Reviewer: I thought that was good though - that were trying to limit the options in the DR?

        successCb: SuccessCallback, errorCb):

    cursor = get_my_connection().cursor()

    # Params for query.
    my_relationship_type: str = get_my_relationship_type(relationship)
    entity_ref_strs = (str(ref) for ref in batch)

    # Run the query.
    placeholders = ", ".join(["?"] * len(batch))  # "?, ?, ?, ..."
    cursor.execute(
        f"""
        SELECT FromRef, ToRef FROM Relationship
        WHERE RelationshipType = ?
        AND FromRef IN ({placeholders})
        ORDER BY FromRef
        """, my_relationship_type, *entity_ref_strs)

    # Get the raw results
    result_set = cursor.fetchall()

    # Gather the results for each input ref.
    result_dict = collections.defaultdict(list)
    for result in result_set:
        from_ref = EntityReference(result[0])
        to_ref = EntityReference(result[1])
        result_dict[from_ref].append(to_ref)

    # Call callbacks.
    for idx, from_ref in enumerate(batch):
        successCb(idx, result_dict[from_ref])
```

#### Page token signature

```python
class PageToken(AbstractPageToken):
    def __init__(self, cursor):
        self.cursor = cursor

def relatedEntities(

>     input: EntityReference, relationship, limit=-1, pageToken=None

Reviewer: [consider]
```suggestion
    input: EntityReference, relationship, limit=-1, prevPageToken=None
```

Author: :thinking: Would it be `prev` or is it more like `current` or even `next`?  Since the token represents the page you haven't got yet.  Depends how you think about it.  Perhaps best to leave it generic?

Reviewer: Yeah- maybe thats the Q - what does the token represent!

) -> Tuple[AbstractPageToken, List[EntityReferences]]:

    if pageToken is None:
        my_relationship_type: str = get_my_relationship_type(relationship)

        cursor = get_my_connection().cursor()
        cursor.execute(
            f"""
            SELECT ToRef FROM Relationship
            WHERE RelationshipType = ?
            AND FromRef = ?
            """, my_relationship_type, str(input))

        pageToken = PageToken(cursor)

    if limit <= 0:
        result_set = pageToken.cursor.fetchall()
    else:
        result_set = pageToken.cursor.fetchmany(limit)

    related_refs = [EntityReference(result[0]) for result in result_set]
    return pageToken, related_refs
```

#### Cursor object signature

The design of DB-API cursors have a somewhat different control flow to
the `AbstractPage` example used throughout, and in practice we would
probably like to mirror a standard such as DB-API in any final design.
However, shimming the DB-API cursor to match our contrived and
ill-conceived `AbstractPage` is a useful exercise.

```python

> class Page(AbstractPage):

Reviewer: [fix] Can we align the limit behaviour between the two implementations? As iirir, this version doesn't have a way to define the size of the first page.

ie. pass limit in `relatedEntities` and set the initial page size there?

Author: The `next` function determines the page size applied to the next `get` call.

Reviewer: But if I just call `get` on the cursor im given by the method, it gets everything defeating paging.

Author: In the current wonky design the initial page size is 0, so it'll be `fetchmany(0)` and you'll get nothing.

Having said that, I agree with your assessment in https://github.com/OpenAssetIO/OpenAssetIO/pull/908#discussion_r1198864935 and will redesign the wonky `AbstractPage` to make a bit more sense.

Reviewer: 🙏

    def __init__(self, cursor):
        self.__cursor = cursor
        self.__is_pending = False
        self.__page_size = 0

    def get(self) -> List[EntityReference]:
        self.__is_pending = False
        result_set = self.__cursor.fetchmany(self.__page_size)
        return [
            EntityReference(result) for result[0] in result_set]

>     def next(self, pageSize):

Reviewer: [question] Is this illustrating the additional cursor functionality (ie arbitrarity advancing without fetching) as assuming (its not entirely clear in the docs) `fetchmany` advances the cursor anyway? If so - maybe calling it `skip` or similar may disambiguate the a host needs to call next to get the next page?

Author: Yeah, the whole `AbstractPage` design is a bit wonky and ill-conceived.  

I didn't want to go back and change all the previous sketches, and thought actually the fact that it doesn't conform to DP-API cursor objects provided a useful exercise in adapting a "real" cursor to an awkward design.  

Given that we can't predict the API of back-end cursor objects, I thought that was a valid exercise.

Author: Oh to actually answer your question, yeah this is illustrating advancing without fetching, and indeed `fetchmany` also advances the cursor.

However, `next` _must_ be called by the user to set the page size for the `get` call, so "skip" doesn't really work, since it's only skipping if `get` isn't called.

Reviewer: Yeah - but I think we have to balance confusing a reader of the DR who will consider this as a proposed cursor design.  The one presented would certainly put me off choosing that option.

Author: Agreed, will redesign throughout.

for retrieving a list of versions of a given entity, and retrieving a
list of entities that are "related to" another entity in some way.

Taking the related entities query as an example: this could be sibling
or parent-child relationships (e.g. shot and sequence), or more loose
relationships (e.g. the project file that was executed to generate a
render).

The current API design provides the host with a list of entity
references that correspond to an input entity-relationship pair.

There may be multiple entity-relationship pairs provided as input to
a single API call, allowing the manager plugin to leverage batching
optimizations. Note that, typically, batching primarily addresses
network latency concerns, in addition to database query optimizations.

In the most general case, the host cannot know the size of the list of
related entity references that will be returned to it. This clearly
poses a scalability issue.

Database and web APIs have had to deal with this problem for decades,
and invariably provide some mechanism of returning slices of result sets
or "pages" that can be iterated through, lazily fetching chunks of data

> in a controlled way.

Reviewer 2: [Consider] I realise we no longer (or never did?) mention stability as a problem in the Background. I wonder if we should prime the concept before we start mentioning it in pros and cons, especially as it's a large part of the reason for rejecting offset.

Author: Originally it was mentioned in the Background, since we enumerated the three different paging styles and made brief mention of the stability issues of offset-based solutions.

I've added a little paragraph on the features of an ideal paging solution in a6785840

showing a straightforward mapping of functionality, and demonstrating
the extensibility to support other common cursor API methods (in this
case, skipping over records without the cost of fetching all the
results).

```python
class Cursor(AbstractCursor):
    def __init__(self, cursor):
        self.__cursor = cursor

    def getPage(self, pageSize) -> List[EntityReference]:
        result_set = self.__cursor.fetchmany(pageSize)
        return [
            EntityReference(result) for result[0] in result_set]

    def skip(self, pageSize):
        self.__cursor.scroll(pageSize)

def relatedEntities(
        input: EntityReference, relationship) -> AbstractCursor:
    my_relationship_type: str = get_my_relationship_type(relationship)

    cursor = get_my_connection().cursor()
    cursor.execute(
        f"""
        SELECT ToRef FROM Relationship
        WHERE RelationshipType = ?
        AND FromRef = ?
        """, my_relationship_type, str(input))

    return Cursor(cursor)
```

## Outcome

We opt for a non-batched cursor object design.

Avoiding conflating batching and pagination simplifies manager plugin
implementations.

> A cursor object maximises future extensibility, where new functionality
> can be added without further polluting the `Manager`/`ManagerInterface`
> class.
> 

Reviewer 2: [Consider]
```suggestion
A cursor object maximises future extensibility, where new functionality
can be added without further polluting the `Manager`/`ManagerInterface`
class.

Whilst we may miss out on some theoretical optimization, we have been 
unable to find any real-world examples showing this sort of batched-page
optimization.
```

    # Call callbacks.
    for idx, from_ref in enumerate(batch):
        successCb(idx, result_dict[from_ref])
```

#### Page token signature

The page token design, as envisioned in earlier sketches, is further
elaborated below, as a possible manager implementation using DB-API.

```python
class PageToken(AbstractPageToken):
    def __init__(self, cursor):
        self.cursor = cursor

def relatedEntities(
    input: EntityReference, relationship, limit=-1, pageToken=None
) -> Tuple[AbstractPageToken, List[EntityReferences]]:

    if pageToken is None:
        my_relationship_type: str = get_my_relationship_type(relationship)

        cursor = get_my_connection().cursor()
        cursor.execute(
            f"""
            SELECT ToRef FROM Relationship
            WHERE RelationshipType = ?
            AND FromRef = ?
            """, my_relationship_type, str(input))

        pageToken = PageToken(cursor)

    if limit <= 0:
        result_set = pageToken.cursor.fetchall()
    else:
        result_set = pageToken.cursor.fetchmany(limit)

    related_refs = [EntityReference(result[0]) for result in result_set]
    return pageToken, related_refs
```

> #### Page token alternative - command pattern

Author: After offline conversaion, added this alternative sketch

```python
class AbstractPage:

    def get(self) -> List[EntityReferences]:
        # Current page of refs.
        ...

    def next(self, pageSize: int):
        # Get next page.
        ...

    def remaining(self) -> Union[int, None]:
        # None means "unsupported".
        ...
```

#### Pros

- Retrieving the data can be decoupled from traversal of pages, allowing
  for lazy network retrieval of desired pages.
- Additional metadata can be included in the cursor (e.g. remaining
  pages) in an extensible way.
- A mutable, stateful cursor object maximizes opportunities for the
  manager plugin to optimize query caches.

#### Cons

- Manager plugin must implement yet another interface.
- Complexity of binding to multiple target languages.
- A mutable, stateful cursor violates the stateless design principle of
  interactions with manager plugins. This causes complications, e.g.
  thread-safety.
- An immutable cursor object (such that `.next()` returns a new cursor
  object), could increase optimisation complexity for the manager
  implementation, e.g. managing query cache sharing, lifetime and
  updates across multiple "live" cursors.

> - Destruction of a cursor object could be used as a signal to the

Reviewer: [question] Not a pro?

Author: Indeed, whoops.

Author: Fixed in 7d540968

  manager that no further pages are required, allowing e.g. cache
  cleanup.

## Considering batching

Additional considerations arise when considering how batching can be
handled algongside paging. There are three broad strategies

- Split batching from paging, using a different API method for each.
- Page through the results from all the batched input elements in
  lockstep, i.e. whole-batch paging.
- Page through the results of each batched input element independently,
  using individual sub-queries within the whole-batch query, i.e.
  per-element paging.

Each of these batched paging designs combines more or less well with the
various paging options presented above.

Henceforth the offset-based solution is rejected, given the lack of
flexibility afforded to the manager's implementation, resulting in poor

> stability and likely manager overhead of record counting.

Author: Moved to an intermediate Outcome section in 2c1020e3

E.g. Page token paged signature

```python
def relatedEntities(
    input: EntityReference, relationship, limit=-1, pageToken=None
) -> Tuple[AbstractPageToken, List[EntityReferences]]: ...
```

E.g. Cursor object paged signature
```python
class AbstractPage:
    def get(self) -> List[EntityReferences]:
        ...
    def next(self, pageSize):
        ...

def relatedEntities(
    input: EntityReference, relationship) -> AbstractPage: ...
```

#### Pros

- Simpler for manager plugin implementors to reason about (as evidenced
  by ad-hoc discussions with collaborators).
- Hosts should usually have a good idea whether batching or paging is
  more appropriate for the query they are making.
- Manager and host do not need to consider both paging and batching at
  the same time, which is a complex convergence of concepts.

#### Cons

- Precludes batching optimisations by the manager when paging is
  required.
- Requires implementation of two methods by the manager: one for
  batched, one for paged.
- Potential difficulty for some hosts/queries to determine a priori
  which API (paging or batch) is the most optimal choice.
- Having the choice adds a foot-gun for host developers to shoot

>   themselves with by accident.

Reviewer: - Precludes hosts atomically expressing more complex queries that may be required by spreadsheet-type UIs.

Author: Added in 86010fc2

### Batched paging option 2 - Whole-batch paging

E.g. Page token

```python
# Callback taking index of batch element and selected page of results.
SuccessCallback = Callable[[int, List[EntityReference]], None]

def relatedEntities(
    batch: List[EntityReference], relationship,
    successCb: SuccessCallback, errorCb,
    limit=-1, pageToken=None
) -> AbstractPageToken: ...
```

E.g. Cursor object

```python
class AbstractPage:
    def get(self) -> List[List[EntityReference]]:
        ...
    def next(self, pageSize):
        ...

def relatedEntities(
    batch: List[EntityReference], relationship) -> AbstractPage: ...
```

#### Pros

- Whole batch is paged at once, retaining batch-awareness in the
  manager for subsequent pages.
- Conceptually simpler API for both host and manager if only a single
  page is active across the whole batch.

#### Cons

- Wasted processing of empty lists by the host when pages are exhausted
  i.e. when there are still pages remaining for some inputs but not
  others.
- Batch elements cannot enter an error state in isolation, since paging
  is whole-batch.

> 

Reviewer: [fix] Something about how no one manager side could figure out in their heads out how you'd actually implement this

E.g. Page token

```python
# Callback taking index of batch element and selected page of results.
SuccessCallback = Callable[[int, List[EntityReference]], None]

def relatedEntities(
    batch: List[EntityReference],
    relationship,
    successCb: SuccessCallback,
    errorCb,
    limit: -1,
    pageTokens: List[AbstractPageToken]
): -> List[AbstractPageToken]: ...
```

E.g. Cursor object

```python
class AbstractPage:
    def get(self) -> List[EntityReferences]:
        ...
    def next(self, pageSize):
        ...

# Callback taking index of batch element and cursor object.
SuccessCallback = Callable[[int, AbstractPage], None]

def relatedEntities(
    batch: List[EntityReference],
    relationship,
    successCb: SuccessCallback,
    errorCb): ...
```

#### Pros

- Manager is under no obligation to consider errors in other streams
  when implementing query behaviour, as pagination method need not be synchronized

#### Cons

- Batching is very difficult to optimize for the manager, especially
  after the first page.
- Batch-optimized queries have unpredictable performance profile for
  the host, given the wide range of potential implementations.

> 

Reviewer: [fix] The host cant express cursor modifications as a batch operation.

Author: I wonder if pros/cons that are specific to a particular paging strategy should go in a new section further down, and ideally only after we've rejected one or more of the options here, so we can reduce the matrix?

Reviewer: Was just trying to keep it a bit more grokable, we know people get lost in these when there are a lot of sections.

Author: OK that's fine.  Though I'm now struggling to understand what your suggested Con means, please can you elaborate?

Reviewer: Sorry was trying to describe how the cursor-per-element approach precludes the host from getting the next page for multiple cursors simultaneously (as they're distinct methods on each individual cursor) which breaks the batch. If you use tokens, then you can supply multiple tokens at once to the second invocation which keeps the batch-ness intact.

Author: Added a Con in 3619805a

by operating on the batch-aware cursor object.

```python
class AbstractPage:

    def get(self) -> List[List[EntityReferences]]:
        # Current page of refs for each element of input batch.
        ...

    def next(self):
        ...

    def remaining(self) -> Union[List[int], None]:
        # None means "unsupported".
        ...

def relatedEntities(
        batch: List[EntityReference], relationship) -> AbstractPage: ...
```
```python
# Usage:

inputs = [EntityReference(...), ...]
relationship = SomeRelationship.traitsData()

# Get intermediate cursor object.
cursor = relatedEntities(inputs, relationship)
```

```python
# Extract all results.

def extract_all(inputs, cursor):
    relateds = [[] for _ in inputs]

    has_more_pages = True
    while has_more_pages:
        # Perhaps yield here to allow UI updates etc...
        has_more_pages = False
        cursor.next(100)
        pages = cursor.get()
        for page, related in zip(pages, relateds):
            if page:
                has_more_pages = True
                related += page

    return relateds
```

#### Cons

- Additional metadata in the object is limited to whole-batch data.

> ## Outcome

Reviewer: [fix] Great work on the restructure. Can we move this before the appendix with out conclusions to date

- no batch paging
- cursor or token

Author: Have we definitively concluded no batch paging?  I had spent some time trying (mostly failing) to sketch how that would look on the manager side, but if we're confident we've rejected batch paging then I can skip that.

Reviewer: We were strongly advised to keep the two things separate if we can. The only noted 'con' would be that it precludes complex UIs requesting their data in one go. It was noted though that we have no concrete use cases for that yet though, and can always add it later once we have some real world scenarios.

Author: Added an(other) intermediate Outcome section in 3854ec13

The current API design provides the host with a list of entity
references that correspond to an input entity-relationship pair.

There may be multiple entity-relationship pairs provided as input to
a single API call, allowing the manager plugin to leverage batching
optimizations. Note that, typically, batching primarily addresses
network latency concerns, in addition to database query optimizations.

In the most general case, the host cannot know the size of the list of
related entity references that will be returned to it. This clearly
poses a scalability issue.

Database and web APIs have had to deal with this problem for decades,
and invariably provide some mechanism of returning slices of result sets
or "pages" that can be iterated through, lazily fetching chunks of data
in a controlled way.

In an ideal situation, requesting subsequent pages has performance
benefits over repeating the same query twice (or more) with different
range arguments, and yet work is performed lazily as pages are requested
so that no unnecessary processing of unused pages is performed.  Also,
pages should be stable, in that results are not skipped nor appear more
than once on successive pages.

> Ultimately, the implementation must be provided by the manager plugin

Author: After offline conversation, added this clarification in 724e9dc0

authors. The mechanism chosen in this DR can only present an abstract
interface that the plugin must satisfy. As a result, characteristics

> such as page stability and performance cannot be guaranteed, but we can
> at least ensure that the chosen mechanism does not preclude such
> optimisations.

Reviewer 2: [Minor][Consider]
Super subjective thing, but this reads to me almost like we're saying "Yes, managers _should_ be stable but we can't force them to be", when I think in actuality we mean more like "Manager may want to be stable, and we don't want to get in the way of that, whilst we also don't want force it."

Maybe?

```suggestion
such as page stability and performance cannot be guaranteed, and nor
do we desire to place such restrictions upon the implementation. However
we can at least ensure that the chosen mechanism does not preclude such
optimisations.
```

Author: Yeah I like it, will commit your suggestion as-is.

    def skip(self, pageSize: int):
        # Skip over records without fetching.
        ...

    def remaining(self) -> Union[int, None]:
        # None means "unsupported".
        ...
```

#### Pros

- Retrieving the data can be decoupled from traversal of pages, allowing
  for lazy network retrieval of desired pages.
- Additional metadata can be included in the cursor (e.g. remaining
  pages) in an extensible way.
- A mutable, stateful cursor object maximizes opportunities for the
  manager plugin to optimize query caches.
- Destruction of a cursor object could be used as a signal to the
  manager that no further pages are required, allowing e.g. cache
  cleanup.

#### Cons

- Manager plugin must implement yet another interface.
- Complexity of binding to multiple target languages.
- A mutable, stateful cursor violates the stateless design principle of
  interactions with manager plugins. This causes complications, e.g.
  thread-safety.
- An immutable cursor object (such that `.next()` returns a new cursor
  object), could increase optimisation complexity for the manager
  implementation, e.g. managing query cache sharing, lifetime and
  updates across multiple "live" cursors.

> - The temptation to directly wrap a "real" DB cursor object in our

Author: After offline conversation/feedback, added this Con and updated the Outcome slightly in 5d85986d

    relationship,
    successCb: SuccessCallback,
    errorCb): ...
```

#### Pros

- Manager is under no obligation to consider errors in other batch
  elements when implementing query behaviour, as pagination method need
  not be synchronized.

#### Pros (cursor object specific)

- Individual cursors can be incremented independently, potentially
  reducing wasted queries.

#### Cons

- Batching is very difficult to optimize for the manager, especially
  after the first page.
- Batch-optimized queries have unpredictable performance profile for
  the host, given the wide range of potential implementations.

#### Cons (cursor object specific)

- Cannot advance to the next page as a batch operation.

### Outcome

Given the reported difficulty of crafting queries that are both batch
and paging-aware, we opt to go with Option 1 and split the paging API(s)
from the batched API(s).

This then leaves only a batch-only signature, plus a decision between
two remaining paging solutions: (non-batched) page token or cursor
objects.

## Outcome

We opt for a non-batched cursor object design.

Avoiding conflating batching and pagination simplifies manager plugin
implementations.

> Whilst we may miss out on some theoretical performance, we have been

Author: Reorganised slightly in 44831de0

unable to find any real-world examples showing this sort of batched-page
optimization.

A cursor object maximises future extensibility, where new functionality
can be added without further polluting the `Manager`/`ManagerInterface`
class. They also naturally capture arbitrary state that the manager
may wish to store as part of the ongoing query.

Cursor objects also reflect the most common programmatic approach to
paging through the results of a database query, presenting a familiar
interface for plugin authors, and mapping well to common back-end SDKs.

> ## Appendix: Considering manager implementations

Author: After offline conversation, to reduce confusion re. dictating a particular implementation, moved all examples to this Appendix section in 0b9b211c

        return pageToken, related_refs
    else:
        raise NotImplementedError()
```

#### Cursor object signature

The following sketch wraps a DB-API cursor in an `AbstractCursor`,
showing a straightforward mapping of functionality, and demonstrating
the extensibility to support other common cursor API methods (in this
case, skipping over records without the cost of fetching all the
results).

```python
class Cursor(AbstractCursor):
    def __init__(self, cursor):
        self.__cursor = cursor

    def getPage(self, pageSize) -> List[EntityReference]:
        result_set = self.__cursor.fetchmany(pageSize)
        return [
            EntityReference(result) for result[0] in result_set]

    def skip(self, pageSize):
        self.__cursor.scroll(pageSize)

def relatedEntities(
        input: EntityReference, relationship) -> AbstractCursor:
    my_relationship_type: str = get_my_relationship_type(relationship)

    cursor = get_my_connection().cursor()
    cursor.execute(
        f"""
        SELECT ToRef FROM Relationship
        WHERE RelationshipType = ?
        AND FromRef = ?
        """, my_relationship_type, str(input))

    return Cursor(cursor)
```

> ### Using a REST API

Author: After offline conversation, added a sketch using a theoretical REST API in 22c9f811

# DR016 One-to-many result paging

> - **Status:** Proposed

Reviewer: [fix] Can we move this to approved

- **Impact:** Medium
- **Driver:** @feltech
- **Approver:** @elliotcmorris @foundrytom

> - **Contributors:** @feltech @elliotcmorris

Reviewer: [fix] Can we add contributors from the threads?

Author: Do you mean what GitHub refers to as "participants" (shown at the top right) on this PR?

Author: Added GitHub participants in 426f7c93

SuccessCallback = Callable[[int, AbstractCursor], None]

def relatedEntities(
    batch: List[EntityReference],
    relationship,
    successCb: SuccessCallback,
    errorCb): ...
```

#### Pros

- Manager is under no obligation to consider errors in other batch
  elements when implementing query behaviour, as pagination method need
  not be synchronized.

#### Pros (cursor object specific)

- Individual cursors can be incremented independently, potentially
  reducing wasted queries.

#### Cons

- Batching is very difficult to optimize for the manager, especially
  after the first page.
- Batch-optimized queries have unpredictable performance profile for
  the host, given the wide range of potential implementations.

#### Cons (cursor object specific)

- Cannot advance to the next page as a batch operation.

### Outcome

Given the reported difficulty of crafting queries that are both batch
and paging-aware, we opt to go with Option 1 and split the paging API(s)
from the batched API(s).

This then leaves only a batch-only signature, plus a decision between
two remaining paging solutions: (non-batched) page token or cursor
objects.

> ## Outcome

Author: Disambiguated in da871bc3
